streams/test-utils/src/main/java/org/apache/kafka/streams/test/TopologyTestDriver.java

```diff
@@ -0,0 +1,593 @@
+ * consumes the named topic, and will invoke the processor(s) downstream of the source.
+ * If your topology's processors forward messages to sinks, your test can then consume these output messages to verify
+ * they match the expected outcome.
+ * For example, if our topology should have generated 2 messages on {@code output-topic-1} and 1 message on
+ * {@code output-topic-2}, then our test can obtain these messages using the
+ * {@link #readOutput(String, Deserializer, Deserializer)} method:
+ *
+ * <pre>{@code
+ * ProducerRecord<String, String> record1 = driver.readOutput("output-topic-1", strDeserializer, strDeserializer);
+ * ProducerRecord<String, String> record2 = driver.readOutput("output-topic-1", strDeserializer, strDeserializer);
+ * ProducerRecord<String, String> record3 = driver.readOutput("output-topic-2", strDeserializer, strDeserializer);
+ * }</pre>
+ *
+ * Again, our example topology generates messages with string keys and values, so we supply our string deserializer
+ * instance for use on both the keys and values. Your test logic can then verify whether these output records are
+ * correct.
+ * Note, that calling {@link ProducerRecord#equals(Object)} compares all attributes including key, value, timestamp,
+ * topic, partition, and headers.
+ * If you only want to compare key and value (and maybe timestamp), using {@link OutputVerifier} instead of
+ * {@link ProducerRecord#equals(Object)} can simplify your code as you can ignore attributes you are not interested in.
+ * <p>
+ * Note, that calling {@code pipeInput()} will also trigger {@link PunctuationType#STREAM_TIME event-time} base
+ * {@link ProcessorContext#schedule(long, PunctuationType, Punctuator) punctuation} callbacks.
+ * However, you won't trigger {@link PunctuationType#WALL_CLOCK_TIME wall-clock} type punctuations that you must
+ * trigger manually via {@link #advanceWallClockTime(long)}.
+ * <p>
+ * Finally, when completed, make sure your tests {@link #close()} the driver to release all resources and
+ * {@link org.apache.kafka.streams.processor.Processor processors}.
+ *
+ * <h2>Processor state</h2>
+ * <p>
+ * Some processors use Kafka {@link StateStore state storage}, so this driver class provides the generic
+ * {@link #getStateStore(String)} as well as store-type specific methods so that your tests can check the underlying
+ * state store(s) used by your topology's processors.
+ * In our previous example, after we supplied a single input message and checked the three output messages, our test
+ * could also check the key value store to verify the processor correctly added, removed, or updated internal state.
+ * Or, our test might have pre-populated some state <em>before</em> submitting the input message, and verified afterward
+ * that the processor(s) correctly updated the state.
+ *
+ * @see ConsumerRecordFactory
+ * @see OutputVerifier
+ */
+@InterfaceStability.Evolving
+public class TopologyTestDriver {
```

Comment:
As a follow-up PR: we should remove the `KStreamTestDriver` and `ProcessorTopologyTestDriver` and replace them with this public class in our internal unit tests as well. It also helps us to identify if there are any gaps that worth closing (i.e. enhancing the `TopologyTestDriver` in next release).

Reply:
Completely agree.